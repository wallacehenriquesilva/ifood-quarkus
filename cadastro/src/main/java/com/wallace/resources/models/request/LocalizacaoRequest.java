package com.wallace.resources.models.request;

import org.eclipse.microprofile.openapi.annotations.media.Schema;

import javax.validation.constraints.NotNull;

public class LocalizacaoRequest {

    @NotNull
    @Schema(description = "Latitude do restaurante", required = true)
    private Double latitude;

    @NotNull
    @Schema(description = "Longitude do restaurante", required = true)
    private Double longitude;

    public LocalizacaoRequest() {
    }

    public LocalizacaoRequest(Double latitude, Double longitude) {
        this.latitude = latitude;
        this.longitude = longitude;
    }

    public Double getLatitude() {
        return latitude;
    }

    public void setLatitude(Double latitude) {
        this.latitude = latitude;
    }

    public Double getLongitude() {
        return longitude;
    }

    public void setLongitude(Double longitude) {
        this.longitude = longitude;
    }
}
