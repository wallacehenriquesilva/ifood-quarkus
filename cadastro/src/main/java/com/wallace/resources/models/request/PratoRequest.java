package com.wallace.resources.models.request;

import org.eclipse.microprofile.openapi.annotations.media.Schema;

import javax.validation.constraints.NotEmpty;
import javax.validation.constraints.NotNull;
import javax.validation.constraints.Size;
import java.math.BigDecimal;

public class PratoRequest {

    @NotEmpty
    @Size(min = 3, max = 50)
    @Schema(description = "Nome do prato", required = true)
    private String nome;

    @NotEmpty
    @Size(min = 5, max = 100)
    @Schema(description = "Descrição do prato", required = true)
    private String descricao;

    @NotNull
    @Schema(description = "Valor do prato", required = true)
    private BigDecimal preco;

    public PratoRequest() {
    }

    public PratoRequest(String nome, String descricao, BigDecimal preco) {
        this.nome = nome;
        this.descricao = descricao;
        this.preco = preco;
    }

    public String getNome() {
        return nome;
    }

    public void setNome(String nome) {
        this.nome = nome;
    }

    public String getDescricao() {
        return descricao;
    }

    public void setDescricao(String descricao) {
        this.descricao = descricao;
    }

    public BigDecimal getPreco() {
        return preco;
    }

    public void setPreco(BigDecimal preco) {
        this.preco = preco;
    }
}
