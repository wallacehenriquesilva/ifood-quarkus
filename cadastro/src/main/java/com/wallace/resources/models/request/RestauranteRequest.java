package com.wallace.resources.models.request;

import com.wallace.entities.RestauranteEntity;
import com.wallace.resources.models.request.validation.ValidModel;
import com.wallace.resources.models.request.validation.ValidRequestModel;
import org.eclipse.microprofile.openapi.annotations.media.Schema;

import javax.validation.ConstraintValidatorContext;
import javax.validation.constraints.NotEmpty;
import javax.validation.constraints.NotNull;
import javax.validation.constraints.Pattern;
import javax.validation.constraints.Size;

@ValidModel
public class RestauranteRequest implements ValidRequestModel {

    @NotEmpty
    @Pattern(regexp = "^[0-9]{2}\\.[0-9]{3}\\.[0-9]{3}\\/[0-9]{4}-[0-9]{2}$", message = "CPNJ inválido.")
    //Usamos o @Schema para documentar a o swagger
    @Schema(description = "CNPJ do restaurante", required = true)
    private String cnpj;

    @NotEmpty
    @Size(min = 3, max = 50)
    @Schema(description = "Nome do restaurante", required = true)
    private String nome;

    @NotEmpty
    @Size(min = 3, max = 50)
    @Schema(description = "Nome do proprietário do restaurante", required = true)
    private String proprietario;

    @NotNull
    @Schema(description = "Localização do restaurante", required = true)
    private LocalizacaoRequest localizacao;


    public RestauranteRequest() {
    }

    public RestauranteRequest(String proprietario, String cnpj, String nome, LocalizacaoRequest localizacao) {
        this.proprietario = proprietario;
        this.cnpj = cnpj;
        this.nome = nome;
        this.localizacao = localizacao;
    }

    public String getProprietario() {
        return proprietario;
    }

    public void setProprietario(String proprietario) {
        this.proprietario = proprietario;
    }

    public String getCnpj() {
        return cnpj;
    }

    public void setCnpj(String cnpj) {
        this.cnpj = cnpj;
    }

    public String getNome() {
        return nome;
    }

    public void setNome(String nome) {
        this.nome = nome;
    }

    public LocalizacaoRequest getLocalizacao() {
        return localizacao;
    }

    public void setLocalizacao(LocalizacaoRequest localizacao) {
        this.localizacao = localizacao;
    }


    /*
    Criação do validador, que vai validar assim que recebermos uma requisição, se o CNPJ já esta cadastrado.
     */
    @Override
    public boolean isValid(ConstraintValidatorContext validatorContext) {

        if (RestauranteEntity.find("cnpj", cnpj).count() > 0) {
            validatorContext.disableDefaultConstraintViolation();

            validatorContext.buildConstraintViolationWithTemplate("CNPJ já cadastrado")
                    .addPropertyNode("cnpj")
                    .addConstraintViolation();
            return false;
        }
        return true;
    }
}
