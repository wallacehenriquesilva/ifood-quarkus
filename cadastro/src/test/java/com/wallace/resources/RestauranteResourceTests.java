package com.wallace.resources;

import com.github.database.rider.cdi.api.DBRider;
import com.github.database.rider.core.api.configuration.DBUnit;
import com.github.database.rider.core.api.configuration.Orthography;
import com.github.database.rider.core.api.dataset.DataSet;
import com.wallace.configurations.CadastroLifeCycleManager;
import com.wallace.resources.models.request.LocalizacaoRequest;
import com.wallace.resources.models.request.RestauranteRequest;
import com.wallace.utils.TokenUtils;
import io.quarkus.test.common.QuarkusTestResource;
import io.quarkus.test.junit.QuarkusTest;
import io.restassured.RestAssured;
import io.restassured.http.ContentType;
import io.restassured.http.Header;
import io.restassured.specification.RequestSpecification;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import javax.ws.rs.core.Response;

@DBRider
//Para carregar datasets e inseri-los em nosso banco. Ele é feito em cima do DBUnit, então, podemos definir uma estratégia ao DBUnit
@DBUnit(caseInsensitiveStrategy = Orthography.LOWERCASE) //Dizemos para usar tudo como lowercase
@QuarkusTest
@QuarkusTestResource(CadastroLifeCycleManager.class)
        //Chama o lifecycle para subir o container do testsContainers
class RestauranteResourceTests {

    private String token; //Token que será utilizado pelas requisições

    @BeforeEach
    public void createToken() throws Exception {
        //Cria o token utilizando o TokenUtils, e usando o arquivo JWTProprietarioClaims.json como token.
        token = TokenUtils.generateTokenString("/JWTProprietarioClaims.json", null);
    }


    private RequestSpecification given() {
        return RestAssured.given()
                .header(new Header("Authorization", "Bearer " + token))
                .contentType(ContentType.JSON);
    }


    //TODO -> O CONTAINER SOBE NORMAL
    @Test
    @DataSet(value = "datasets/restaurantes-cenario-1.yml", cleanAfter = true)
    /**
     *
     *  @DataSet -> Da para configurarmos varias coisas, como o que sera executado no começo, no fim, etc.
     *  Podemos usar também o @ExpectedDataSet() passando o arquivo do dataset esperado após as ações do teste.
     */
    void getAllRestaurants() {
        final String result = given()
                .when().get("/api/v1/restaurantes")
                .then()
                .statusCode(Response.Status.OK.getStatusCode())
                .extract().asString();
        //Valida o retorno se ele e exatamente igual ao que temos que salvar em um arquivo com o padrao
        //nome da classe . (ponto) nome do metodo . (ponto) approved.json
        Approvals.verifyJson(result);
    }

    @Test
    @DataSet(value = "datasets/restaurantes-cenario-1.yml", cleanAfter = true)
    void getOneRestaurant() {
        final String result = given()
                .when()
                .get("/api/v1/restaurantes/123")
                .then()
                .statusCode(Response.Status.OK.getStatusCode())
                .extract().asString();
        Approvals.verifyJson(result);
    }


    @Test
    @DataSet(value = "datasets/restaurantes-cenario-1.yml", cleanAfter = true)
    void createRestaurant() {
        final LocalizacaoRequest localizacaoRequest = new LocalizacaoRequest(
                12354.55,
                12354.55
        );

        final RestauranteRequest restauranteRequest = new RestauranteRequest(
                "Wallace Silva",
                "12.352.123/1542-20",
                "Restaurante Update",
                localizacaoRequest
        );

        given()
                .when()
                .header("Content-Type", "application/json")
                .body(restauranteRequest)
                .post("/api/v1/restaurantes")
                .then()
                .statusCode(Response.Status.CREATED.getStatusCode());
    }


    @Test
    @DataSet(value = "datasets/restaurantes-cenario-1.yml", cleanAfter = true)
    void updateRestaurant() {
        final LocalizacaoRequest localizacaoRequest = new LocalizacaoRequest(
                12354.55,
                12354.55
        );

        final RestauranteRequest restauranteRequest = new RestauranteRequest(
                "Wallace Silva",
                "12.352.123/1542-20",
                "Restaurante Update",
                localizacaoRequest
        );

        given()
                .when()
                .header("Content-Type", "application/json")
                .body(restauranteRequest)
                .put("/api/v1/restaurantes/123")
                .then()
                .statusCode(Response.Status.NO_CONTENT.getStatusCode());
    }

    @Test
    @DataSet(value = "datasets/restaurantes-cenario-1.yml", cleanAfter = true)
    void deleteRestaurant() {
        given()
                .when()
                .header("Content-Type", "application/json")
                .delete("/api/v1/restaurantes/123")
                .then()
                .statusCode(Response.Status.NO_CONTENT.getStatusCode());
    }

    //TODO -> FAZER TESTES DE ERRO


}
