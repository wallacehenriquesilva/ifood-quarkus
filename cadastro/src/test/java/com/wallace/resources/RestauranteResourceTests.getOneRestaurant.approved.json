{
  "id": 123,
  "nome": "Restaurante 1",
  "proprietario": "Wallace"
}